src/bsoncxx/types/value.cpp

```diff
@@ -84,8 +88,12 @@ value::~value() {
     }
 }
 
-bsoncxx::type value::type() const {
-    return _type;
+stdx::optional<bsoncxx::type> value::type() const {
+    if (_type == 0) {
```

Comment:
May not be able to do so if the types are different.

However, maybe the following works?

return _type ? { static_cast<bsoncxx::type>(_type); } : { stdx::nullopt };